Add support for asynchronous event handlers

marmarek opened the pull request:
Some event handlers need to call async actions (like qrexec service). Add support for it.
Two important design choices:
 - event handlers can still be normal python function (instead of a coroutine), if given handler don't need any async feature
 - not all events support async handlers - see appropriate event documentation for this information

woju commented on qubes/events.py:
```diff
@@ -157,50 +160,92 @@ def _fire_event_in_order(self, order, event, kwargs):
                     key=(lambda handler: hasattr(handler, 'ha_bound')),
                     reverse=True):
                 effect = func(self, event, **kwargs)
+                # "is True" to rule out unittest.mock.Mock...
```
#wat?

marmarek replied:
If func is unittest.mock.Mock instance, it do have attribute `ha_async` (in fact: any attribute), which is a new unittest.mock.Mock instance, `bool(unittest.mock.Mock())` is true

woju replied:
Please no. Fix your mock instead.

marmarek replied:
This will be gone after switching to iscoroutinefunction.

woju replied:
Ah, OK.

woju commented on qubes/events.py:
```diff
@@ -157,50 +160,92 @@ def _fire_event_in_order(self, order, event, kwargs):
+        if pre_event:
+            order = itertools.chain((self,), self.__class__.__mro__)
+        else:
+            order = itertools.chain(reversed(self.__class__.__mro__), (self,))
```
```python
order = itertools.chain((self,), self.__class__.__mro__)
if not pre_event:
  order = reversed(order)
```
Same as the last one: less places to err.

Also, I'd rename `_fire_event_in_order` to `_fire_event` and put the above logic in there, since it is shared between this and and `fire_event_async` (c.f. less places to err). And I'd inline `_handle_event_async` in the latter.

marmarek replied:
Doesn't work:
```python
>>> import itertools
>>> order = itertools.chain((1, ), (2, 3, 4))
>>> reversed(order)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: argument to reversed() must be a sequence
```
Should I force it into a list or something? Or leave the old version (but still moved to the other place)?

woju replied:
Sorry, my fault. `order = reversed(list(order))`.

woju commented on qubes/events.py:
```diff
@@ -157,50 +160,92 @@ def _fire_event_in_order(self, order, event, kwargs):
                 effect = func(self, event, **kwargs)
+                # "is True" to rule out unittest.mock.Mock...
+                if getattr(func, 'ha_async', False) is True:
+                    assert effect is not None
```
I'd very much prefer `assert inspect.isawaitable(effect)` instead of `is not None`. `*.iscoroutine()` may not be enough, since someone might like to return a plain `Future` (not a `Task`) or whatever.

marmarek replied:
AFAIR there was some problem with `inspect.isawaitable()` - we did considered it for very similar case in storage, but discarded it there.

woju replied:
```python
def isawaitable(object):
    """Return true if object can be passed to an ``await`` expression."""
    return (isinstance(object, types.CoroutineType) or
            isinstance(object, types.GeneratorType) and
                bool(object.gi_code.co_flags & CO_ITERABLE_COROUTINE) or
            isinstance(object, collections.abc.Awaitable))
```

What's wrong with it?

marmarek replied:
To be honest, I don't remember. I'll try ;)

woju commented on qubes/events.py:
```diff
@@ -157,50 +160,92 @@ def _fire_event_in_order(self, order, event, kwargs):
-        Handlers are called for class and all parent classess, in **true**
-        method resolution order. This is intended for ``-pre-`` events, where
-        order of invocation should be reversed.
+        Thi method call both synchronous and asynchronous handlers. Order of
```
s/Thi/This/

woju commented on qubes/events.py:
```diff
@@ -157,50 +160,92 @@ def _fire_event_in_order(self, order, event, kwargs):
                     reverse=True):
                 effect = func(self, event, **kwargs)
+                # "is True" to rule out unittest.mock.Mock...
+                if getattr(func, 'ha_async', False) is True:
```
Now I think, couldn't we replace this one with `asyncio.iscoroutinefunction(func)` (not to be confused with `inspect.iscoroutinefunction()`, which is `False` on `@asyncio.coroutine`s)?

That way we'd deal with one corner case:
```python
class MyClass(qubes.events.Emitter):
  @qubes.events.handler('my-event', is_async=True)
  # NOT asyncio.coroutine
  def on_my_event(self, event):
    return asyncio.Future()
```

With `iscoroutinefunction`, the above example, admittedly valid but unreadable, becomes forbidden. And we wouldn't have to remember about `is_async=` and keep it in sync with reality.

Argument against: there are 2 incompatible versions of `iscoroutinefunction()` in python's stdlib (the difference is around `@asyncio.coroutine` as above) and 2.5 versions of `iscoroutine()` (there is also `isinstance(obj, types.CoroutineType)`,which is same as `inspect.iscoroutine()` and not compatible with `asyncio.iscoroutine()`). We can get it right, but after that, we are in the area of "what could possibly go wrong".

marmarek replied:
That wouldn't work on your hacky coroutines which do not have `@asyncio.coroutine`, but have a single `return some_other_coroutine(...)`.

marmarek replied:
That would also solve the problem with `emitter.add_handler()`. 👍 for removing those hacky coroutines without `@asyncio.coroutine`.

woju commented on qubes/ext/__init__.py:
```diff
@@ -86,6 +88,9 @@ def decorator(func):
 
+        if kwargs.get('is_async', False):
+            func.ha_async = True
+
```
`kwargs.get()` in that function was a workaround to get keyword-only arguments in python2 (this is one of the oldest parts of core3).

#TODO rewrite as keyword-only argument. Low priority.

marmarek commented:
extracted from commit message:
> One limitation (for now) is that Emiter.add_handler does not support
registering async handlers (`func.ha_async = True` can raise for example
`AttributeError: 'method' object has no attribute 'ha_async'`).

Is it ok to leave if this way, or there is some _simple_ method to solve it?

marmarek commented:
Force-pushed fixed version. Oh, it fails on travis, but works on my machine. The fail is about `inspect.isawaitable()`...

marmarek commented:
So, back to `assert effect is not None`? Or any better idea?

marmarek commented:
Dropped that assert (it was to detect non-coroutines registered with is_async=True, so it is irrelevant now).
And now fails on https://bugs.python.org/issue28703 ("Fix asyncio.iscoroutinefunction to handle Mock objects"), fixed in python 3.5.3 (we have 3.5.2 on travis...).
And lets see:
```patch
-    return (getattr(func, '_is_coroutine', False) or
+    return (getattr(func, '_is_coroutine', None) is _is_coroutine or
```
So, exactly the syntax you didn't liked 😛 
Added a workaround in tests...

woju commented on doc/qubes-events.rst:
```diff
@@ -143,6 +143,75 @@ returned to the caller as list. The order of this list is undefined.
+Event handlers can be defined as coroutine. This way they can execute long
+running actions without blocking the whole qubesd process. To define
+asynchronous event handler, annotate a coroutine (a function defined with
+`async def`, or decorated with `py:func:`asyncio.coroutine`) with
```
this should be
```rst
:py:func:`asyncio.coroutine`
```

see the problem with syntax highlighting on github

marmarek commented:
@woju I need you to approve this again, since I've pushed new commit.
